### ISSUE TITLE
destructuring assignment with new name can break mangled

### ISSUE THREAD

**Issue by srguiwiz** [javascript-obfuscator/javascript-obfuscator#355]
## Expected Behavior

Resulting code to work correctly.

## Current Behavior

Error message when running code says "Identifier 'c' has already been declared".

## Steps to Reproduce (for bugs)

At https://obfuscator.io/ on 2018-11-20:

Paste this code
```
var o = { a: 11, b: 12, c: 13, d: 14, e: 15 };
function f(obj) {
  const { a: bananas, b: apples, c: d, d: c, e: eggplant } = obj; // destructuring with new names, a swap
  console.log("in function", obj, apples, bananas, c, d, eggplant );
  return { apples, bananas, c, d, eggplant };
}
var o2 = f(o);
var { apples, bananas, c: cantaloupe, d: date, eggplant } = o2;
console.log("returned from function", o2, apples, bananas, cantaloupe, date, eggplant );
```
Have all obfuscator features off.  Have Identifier Names Generator on **mangled**.  Run (evaluate) the code when obfuscated.  Get
```
Identifier 'c' has already been declared
```

Reason is clear in output, **obj has been renamed c**, and new name c in **const { d: c }** conflicts with that.
```
var o = {
    'a': 0xb,
    'b': 0xc,
    'c': 0xd,
    'd': 0xe,
    'e': 0xf
};
function f(c) { // THIS RENAMED PARAMETER CONFLICTS
    const {
        a: bananas,
        b: apples,
        c: d,
        d: c, // CONFLICTS WITH THIS UNCHANGED USER CODE
        e: eggplant
    } = c;
    console['log']('in\x20function', c, apples, bananas, c, d, eggplant);
    return {
        'apples': apples,
        'bananas': bananas,
        'c': c,
        'd': d,
        'eggplant': eggplant
    };
}
var o2 = f(o);
var {
    apples,
    bananas,
    c: cantaloupe,
    d: date,
    eggplant
} = o2;
console['log']('returned\x20from\x20function', o2, apples, bananas, cantaloupe, date, eggplant);
```

This code runs fine in Node and Chrome and Firefox.

This sample took a while to construct.  Use is where code is a mix of hand coded and generated.  In that environment, short variable names may be generated, and not under direct control of a programmer.

A solution seems to be to correctly recognize destructuring renaming and avoid a conflict.

Note this problem **does not occur if not destructuring**:
```
var o = { a: 11, b: 12, c: 13, d: 14, e: 15 };
function f(obj) {
  const bananas = obj.a, apples = obj.b, d = obj.c, c = obj.d, eggplant = obj.e; // no destructuring
  console.log("in function", obj, apples, bananas, c, d, eggplant );
  return { apples, bananas, c, d, eggplant };
}
var o2 = f(o);
var { apples, bananas, c: cantaloupe, d: date, eggplant } = o2;
console.log("returned from function", o2, apples, bananas, cantaloupe, date, eggplant );
```
because then it nicely avoids conflicts when renaming
```
function f(c) {
    const d = c['a'], e = c['b'], f = c['c'], g = c['d'], h = c['e'];
```

Another workaround is to have Identifier Names Generator on **hexdecimal**.  But, that produces bigger code.

## Your Environment

* Obfuscator version used: 0.18.1
* Node version used: 10.x

## Minimal working example that will help to reproduce issue

As pasted above,
```
var o = { a: 11, b: 12, c: 13, d: 14, e: 15 };
function f(obj) {
  const { a: bananas, b: apples, c: d, d: c, e: eggplant } = obj; // use case involves swaps
  console.log("in function", obj, apples, bananas, c, d, eggplant );
  return { apples, bananas, c, d, eggplant };
}
var o2 = f(o);
var { apples, bananas, c: cantaloupe, d: date, eggplant } = o2;
console.log("returned from function", o2, apples, bananas, cantaloupe, date, eggplant );
```

**Comment by zamotkin** [comment]
I minimized your example to:
```
function f(obj) {
    const {c} = obj;
}
```

and resolved this issue with #441. There are no collisions with non-replaceable and generated variable names now.

**Comment by sanex3339** [comment]
Fixed in `0.18.6` version
